Use a C-safe return type for `__rust_[ui]128_*` overflowing intrinsics

tgross35 opened the pull request:
Combined with [1], this will change the overflowing multiplication operations to return an `extern "C"`-safe type.

Link: https://github.com/rust-lang/compiler-builtins/pull/735 [1]

tgross35 committed:
Use a C-safe return type for `__rust_[ui]128_*` overflowing intrinsics

Combined with [1], this will change the overflowing multiplication
operations to return an `extern "C"`-safe type.

Link: https://github.com/rust-lang/compiler-builtins/pull/735 [1]

tgross35 commented:
r? @bjorn3 @antoyo 

There are a couple other places I need to update for GCC still.

tgross35 commented:
For cranelift, is there a way to load the stack slot so it can be combined with the integer for a `CValue::by_val_pair` @bjorn3?

bjorn3 commented:
You can use `oflow_out_place.to_cvalue(fx).load_scalar(fx)`.

bjorn3 commented on compiler/rustc_codegen_cranelift/src/compiler_builtins.rs:
```diff
@@ -43,7 +43,7 @@ builtin_functions! {
     fn __umodti3(n: u128, d: u128) -> u128;
     fn __modti3(n: i128, d: i128) -> i128;
-    fn __rust_u128_mulo(a: u128, b: u128) -> (u128, bool);
+    fn __rust_u128_mulo(a: u128, b: u128, oflow: &mut bool) -> u128;
```
In the compiler-builtins PR you are using `&mut i32` instead of `&mut bool`.

tgross35 replied:
Thanks. I had `bool` originally but changed to `i32` for consistency with the C APIs.

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/12382255892/job/34562601106) [(plain)](https://github.com/rust-lang/rust/commit/c2a51c9c30284d2dfe698f7872beaa2b30f1215d/checks/34562601106/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#13 2.706 Building wheels for collected packages: reuse
#13 2.707   Building wheel for reuse (pyproject.toml): started
#13 2.956   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 2.956   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132720 sha256=026f3bb0f1aa8090b861fd0a0939cb1a782396d84c8aab7875096557d637a0f6
#13 2.957   Stored in directory: /tmp/pip-ephem-wheel-cache-580rynes/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#13 2.959 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#13 3.353 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#13 3.354 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 3.875 Collecting virtualenv
#13 3.875 Collecting virtualenv
#13 3.912   Downloading virtualenv-20.28.0-py3-none-any.whl (4.3 MB)
#13 3.983      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.3/4.3 MB 62.5 MB/s eta 0:00:00
#13 4.036 Collecting platformdirs<5,>=3.9.1
#13 4.039   Downloading platformdirs-4.3.6-py3-none-any.whl (18 kB)
#13 4.074 Collecting filelock<4,>=3.12.2
#13 4.094 Collecting distlib<1,>=0.3.7
#13 4.097   Downloading distlib-0.3.9-py2.py3-none-any.whl (468 kB)
#13 4.105      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 469.0/469.0 KB 77.8 MB/s eta 0:00:00
#13 4.105      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 469.0/469.0 KB 77.8 MB/s eta 0:00:00
#13 4.186 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 4.368 Successfully installed distlib-0.3.9 filelock-3.16.1 platformdirs-4.3.6 virtualenv-20.28.0
#13 DONE 4.4s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      128960 kB
DirectMap2M:     5113856 kB
DirectMap1G:    13631488 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
WARN: currently no CI rustc builds have rustc debug assertions enabled. Please either set `rust.debug-assertions` to `false` if you want to use download CI rustc or set `rust.download-rustc` to `false`.
downloading https://static.rust-lang.org/dist/2024-11-27/rustfmt-nightly-x86_64-unknown-linux-gnu.tar.xz
---
##[endgroup]
fmt check
fmt: checked 5732 files
tidy check
tidy error: invalid source: "git+https://github.com/tgross35/compiler-builtins.git?branch=overflowing-c-safe-ret#8cb74d05a80a5782a5377a825d2ddd7b950bf220"
tidy error: invalid source: "git+https://github.com/tgross35/compiler-builtins.git?branch=overflowing-c-safe-ret#8cb74d05a80a5782a5377a825d2ddd7b950bf220"
tidy error: invalid license `Zlib` in `registry+https://github.com/rust-lang/crates.io-index#foldhash@0.1.3`
Go to `src/tools/tidy/src/deps.rs:230` for the list.
tidy error: Dependency for compiler/rustc_codegen_cranelift not explicitly permitted: registry+https://github.com/rust-lang/crates.io-index#foldhash@0.1.3
removing old virtual environment
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (24.3.1)
All checks passed!
checking C++ file formatting
some tidy checks failed
some tidy checks failed
Command has failed. Rerun with -v to see more details.
  local time: Tue Dec 17 22:18:57 UTC 2024
  network time: Tue, 17 Dec 2024 22:18:57 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

bors commented:
:umbrella: The latest upstream changes (presumably #134844) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bjorn3 commented on compiler/rustc_codegen_cranelift/Cargo.toml:
```diff
@@ -23,7 +23,8 @@ libloading = { version = "0.8.0", optional = true }
 [patch.crates-io]
-# Uncomment to use local checkout of cranelift
+# todo: remove patch
+compiler_builtins = { git = "https://github.com/tgross35/compiler-builtins.git", package = "compiler_builtins", branch = "overflowing-c-safe-ret" }
```
This doesn't apply to building the sysroot. Also the patch in library/Cargo.toml should be enough to apply while building a sysroot for cg_clif.

bjorn3 commented on compiler/rustc_codegen_cranelift/Cargo.lock:
```diff
@@ -16,15 +16,15 @@ dependencies = [
+version = "1.0.94"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "b3d1d046238990b9cf5bcde22a3fb3584ee5cf65fb2765f454ed428c7a0063da"
+checksum = "c1fd03a028ef38ba2276dce7e33fcd6369c158a1bca17946c4b1b701891c1ff7"
```
The lockfile changes should be reverted.

tgross35 replied:
Yeah, the entire second commit is just for testing and will be dropped. My plan is once you approve the Cranelift changes and somebody from GCC has looked that portion over, I will merge https://github.com/rust-lang/compiler-builtins/pull/735, drop the second commit (which add my crates-io` patch), replace it with a normal `compiler-builtins` bump, and then this PR's merge will change everything atomically.

However, I am stuck completing the current bump at https://github.com/rust-lang/rust/pull/135180, so I will hold off until that gets fixed.

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/12698277021/job/35396414420) [(plain)](https://github.com/rust-lang/rust/commit/7e67edf978dd6f907fbe58a0e36898512c36942a/checks/35396414420/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
curl: (22) The requested URL returned error: 404
curl: (22) The requested URL returned error: 404
ERROR: failed to download llvm from ci

    HELP: There could be two reasons behind this:
        1) The host triple is not supported for `download-ci-llvm`.
        2) Old builds get deleted after a certain time.
    HELP: In either case, disable `download-ci-llvm` in your config.toml:
    [llvm]
    download-ci-llvm = false
    
Build completed unsuccessfully in 0:00:22
```

</details>

tgross35 commented:
@rust-lang/wg-gcc-backend would one of you mind taking a look at the GCC changes? I am assuming that since this bump means the intrinsics are changing to get the same signature (type aside) as `__mulosi4` they no longer need to be special cased. However, I am not very familiar with this code or whether it gets tested in CI somehow.

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/12701196739/job/35405284955) [(plain)](https://github.com/rust-lang/rust/commit/3c4707e58315fb8e98cd2fdf4dc31ee3cfd77a61/checks/35405284955/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#13 2.808 Building wheels for collected packages: reuse
#13 2.809   Building wheel for reuse (pyproject.toml): started
#13 3.015   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 3.016   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132718 sha256=f1d7d99e2187cfcc743e73878997db8829d4c560719802a1230bd160f7d9cfca
#13 3.016   Stored in directory: /tmp/pip-ephem-wheel-cache-c0v7wpz0/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#13 3.019 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#13 3.418 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#13 3.418 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 3.948 Collecting virtualenv
#13 3.948 Collecting virtualenv
#13 4.019   Downloading virtualenv-20.28.1-py3-none-any.whl (4.3 MB)
#13 4.127      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.3/4.3 MB 40.9 MB/s eta 0:00:00
#13 4.182 Collecting platformdirs<5,>=3.9.1
#13 4.196   Downloading platformdirs-4.3.6-py3-none-any.whl (18 kB)
#13 4.232 Collecting filelock<4,>=3.12.2
#13 4.265 Collecting distlib<1,>=0.3.7
#13 4.279   Downloading distlib-0.3.9-py2.py3-none-any.whl (468 kB)
#13 4.286      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 469.0/469.0 KB 89.6 MB/s eta 0:00:00
#13 4.286      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 469.0/469.0 KB 89.6 MB/s eta 0:00:00
#13 4.367 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 4.549 Successfully installed distlib-0.3.9 filelock-3.16.1 platformdirs-4.3.6 virtualenv-20.28.1
#13 DONE 4.7s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      100288 kB
DirectMap2M:     6191104 kB
DirectMap1G:    12582912 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
WARN: currently no CI rustc builds have rustc debug assertions enabled. Please either set `rust.debug-assertions` to `false` if you want to use download CI rustc or set `rust.download-rustc` to `false`.
downloading https://static.rust-lang.org/dist/2025-01-08/rustfmt-nightly-x86_64-unknown-linux-gnu.tar.xz
---
##[endgroup]
fmt check
fmt: checked 5763 files
tidy check
tidy error: invalid source: "git+https://github.com/tgross35/compiler-builtins.git?branch=overflowing-c-safe-ret#8cb74d05a80a5782a5377a825d2ddd7b950bf220"
tidy error: invalid source: "git+https://github.com/tgross35/compiler-builtins.git?branch=overflowing-c-safe-ret#8cb74d05a80a5782a5377a825d2ddd7b950bf220"
removing old virtual environment
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (24.3.1)
All checks passed!
checking python file formatting
28 files already formatted
checking C++ file formatting
checking C++ file formatting
some tidy checks failed
Command has failed. Rerun with -v to see more details.
  local time: Fri Jan 10 01:04:41 UTC 2025
  network time: Fri, 10 Jan 2025 01:04:41 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

antoyo commented on compiler/rustc_codegen_gcc/src/int.rs:
```diff
@@ -322,34 +322,23 @@ impl<'a, 'gcc, 'tcx> Builder<'a, 'gcc, 'tcx> {
-                            _ => unreachable!(),
-                        },
-                    };
-                    return self.operation_with_overflow(func_name, lhs, rhs);
```
From what I can see, this change doesn't look right as this will fall to calling `self.overflow_call()` below which expects that the overflow flag is returned from the function call (like [these GCC intrinsics](https://gcc.gnu.org/onlinedocs/gcc/Integer-Overflow-Builtins.html)) whereas [your change](https://github.com/rust-lang/compiler-builtins/pull/735/files#diff-ef76691bf5a204c8bd0ad9edd0fbaccf3f0284936996f0183ce49140aab10457L69) seems to return the integer result and not the overflow flag.

My guess is that you'll need to keep this special handling and change the method `operation_with_overflow` accordingly as these intrinsics are also used in `src/intrinsic/mod.rs`.

AFAIK, the only related tests that run in this CI are [those lines](https://github.com/rust-lang/rust/blob/master/compiler/rustc_codegen_gcc/example/std_example.rs#L61-L74).

tgross35 replied:
Thanks for taking a look. Doesn't this block currently [sometimes use `__mulosi4`](https://github.com/rust-lang/rust/blob/824759493246ee383beb9cd5ceffa0e15deb9fa4/compiler/rustc_codegen_gcc/src/int.rs#L346-L363) without an early turn? After the change to `compiler_builtins`, `__rust_*128_o` should be consistent with the [`__mulo*i4` functions](https://github.com/rust-lang/compiler-builtins/blob/ba84ef86cc56f6f3404c0b163da419146e0b0fa0/src/int/mul.rs#L112-L125).

> AFAIK, the only related tests that run in this CI are [those lines](https://github.com/rust-lang/rust/blob/master/compiler/rustc_codegen_gcc/example/std_example.rs#L61-L74).

Is there any kind of codegen/asm test that should be added here? Or is the existing test sufficient once CI gets that far.

antoyo replied:
> Thanks for taking a look. Doesn't this block currently [sometimes use `__mulosi4`](https://github.com/rust-lang/rust/blob/824759493246ee383beb9cd5ceffa0e15deb9fa4/compiler/rustc_codegen_gcc/src/int.rs#L346-L363) without an early turn? After the change to `compiler_builtins`, `__rust_*128_o` should be consistent with the [`__mulo*i4` functions](https://github.com/rust-lang/compiler-builtins/blob/ba84ef86cc56f6f3404c0b163da419146e0b0fa0/src/int/mul.rs#L112-L125).

Indeed, it looks like there might be a bug in here.

> Is there any kind of codegen/asm test that should be added here? Or is the existing test sufficient once CI gets that far.

What do you mean? There's more tests that are ran in the repo of rustc_codegen_gcc itself.

tgross35 replied:
> Indeed, it looks like there might be a bug in here.

I changed this so `mulo` and the `__rust` functions should now be taking the same codepath, still have to update the rest.

> > Is there any kind of codegen/asm test that should be added here? Or is the existing test sufficient once CI gets that far.
> 
> What do you mean? There's more tests that are ran in the repo of rustc_codegen_gcc itself.

I am just wondering whether I should add another test anywhere to verify the new behavior, especially if `__mulosi4` may have been called incorrectly before. And if so, what format/location this should be (for LLVM I'd probably add a codegen test, but I'm not sure what the equivalent would be here).

antoyo replied:
I was expecting that there would be UI tests for these, so that every backend could benefit from them.
Do they have to be codegen tests?

There are integer tests [here](https://github.com/rust-lang/rust/blob/master/compiler/rustc_codegen_gcc/tests/run/int.rs) for the GCC codegen, but they are not ran in the Rust CI, as far as I know, so if it's too much of a pain, we can open [an issue](https://github.com/rust-lang/rustc_codegen_gcc/issues) to remind me to add tests for this in the future.

tgross35 replied:
Looking at https://rust.godbolt.org/z/eWfPK6rGW it seems like GCC lowers these operations to assembly (LLVM as well), which possibly explains why the test was not failing before even if the `__mulosi4` call had a bug. I am not very familiar with this backend but I take it GCC intercepts the call to `__builtin_sub_overflow` and adjusts for the correct size?

If it doesn't emit this call then I am not sure what kind of test would work here, but can add something if you have any ideas. I updated the code and it should be ready for review, it builds but have not been able to run tests.

> I was expecting that there would be UI tests for these, so that every backend could benefit from them.

Agreed that we should have something end-to-end here, possibly reusing the test generators from compiler-builtins/libm. I think the status quo assumption is that we don't need to test the lowering too heavily since that is the backend's responsibility, but it would still be good to have something systematic.

antoyo replied:
> I am not very familiar with this backend but I take it GCC intercepts the call to `__builtin_sub_overflow` and adjusts for the correct size?

Yes. [The doc](https://gcc.gnu.org/onlinedocs/gcc/Integer-Overflow-Builtins.html) mentions:

> The first built-in function allows arbitrary integral types for operands and the result type must be pointer to some integral type other than enumerated or boolean type, the rest of the built-in functions have explicit integer types.

I reviewed your code: it looks good, thanks!

[This test](https://github.com/rust-lang/rust/blob/master/compiler/rustc_codegen_gcc/tests/run/int.rs#L214-L218) does call `overflowing_mul`, so I'm not sure why it wasn't tested there.
Any ideas?

antoyo replied:
Yes, that CI job uses a GCC where 128-bit integers are disabled to test this indeed.

Right, that must be why it wasn't tested. But I think `i64` is supported on all platforms, so I'll open an issue to check that.

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/12701828475/job/35407019751) [(plain)](https://github.com/rust-lang/rust/commit/5497d75b63dde2d13585b7f60adbbd7427bab2b8/checks/35407019751/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#13 2.910 Building wheels for collected packages: reuse
#13 2.911   Building wheel for reuse (pyproject.toml): started
#13 3.122   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 3.123   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132718 sha256=f1d7d99e2187cfcc743e73878997db8829d4c560719802a1230bd160f7d9cfca
#13 3.123   Stored in directory: /tmp/pip-ephem-wheel-cache-9p75tdz0/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#13 3.126 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#13 3.528 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#13 3.528 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 4.062 Collecting virtualenv
#13 4.062 Collecting virtualenv
#13 4.108   Downloading virtualenv-20.28.1-py3-none-any.whl (4.3 MB)
#13 4.173      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.3/4.3 MB 67.9 MB/s eta 0:00:00
#13 4.228 Collecting filelock<4,>=3.12.2
#13 4.231   Downloading filelock-3.16.1-py3-none-any.whl (16 kB)
#13 4.248 Collecting distlib<1,>=0.3.7
#13 4.251   Downloading distlib-0.3.9-py2.py3-none-any.whl (468 kB)
#13 4.259      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 469.0/469.0 KB 81.6 MB/s eta 0:00:00
#13 4.291 Collecting platformdirs<5,>=3.9.1
#13 4.294   Downloading platformdirs-4.3.6-py3-none-any.whl (18 kB)
#13 4.376 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 4.560 Successfully installed distlib-0.3.9 filelock-3.16.1 platformdirs-4.3.6 virtualenv-20.28.1
#13 DONE 4.6s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      131008 kB
DirectMap2M:     6160384 kB
DirectMap1G:    12582912 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
WARN: currently no CI rustc builds have rustc debug assertions enabled. Please either set `rust.debug-assertions` to `false` if you want to use download CI rustc or set `rust.download-rustc` to `false`.
downloading https://static.rust-lang.org/dist/2025-01-08/rustfmt-nightly-x86_64-unknown-linux-gnu.tar.xz
---
##[endgroup]
fmt check
fmt: checked 5763 files
tidy check
tidy error: invalid source: "git+https://github.com/tgross35/compiler-builtins.git?branch=overflowing-c-safe-ret#8cb74d05a80a5782a5377a825d2ddd7b950bf220"
tidy error: invalid source: "git+https://github.com/tgross35/compiler-builtins.git?branch=overflowing-c-safe-ret#8cb74d05a80a5782a5377a825d2ddd7b950bf220"
removing old virtual environment
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (24.3.1)
All checks passed!
checking python file formatting
28 files already formatted
checking C++ file formatting
checking C++ file formatting
some tidy checks failed
Command has failed. Rerun with -v to see more details.
  local time: Fri Jan 10 02:06:35 UTC 2025
  network time: Fri, 10 Jan 2025 02:06:35 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

tgross35 commented:
@bjorn3 the is segfaulting at [this line](https://github.com/rust-lang/rust/blob/0799cbf2cac49fa38af9b25bbe433920bc5c9532/compiler/rustc_codegen_cranelift/example/std_example.rs#L76) within `__rust_i128_addo`. The parameters look fine but it tries to `mov    %rcx,(%rdi)` and store the oflow bool to the first argument.

```text
rax            0x377cf7d5a29495da  3998343066125899226                                                                                                                                                         rbx            0x0                 0                                                                                                                                                                           rcx            0x0                 0                                                                                                                                                                           rdx            0xa                 10                                                                                                                                                                          rsi            0x5b232dc60cb0dfd   410446438142250493                                                                                                                                                          rdi            0x377cf7d5a29495da  3998343066125899226                                                                                                                                                         rbp            0x7fffffffdb00      0x7fffffffdb00                                                                                                                                                              rsp            0x7fffffffc9f0      0x7fffffffc9f0                                                                                                                                                              r8             0x7fffffffda90      140737488345744                                                                                                                                                             r9             0x5555555fefe0      93824992931808 
```

tgross35 commented:
Actually that would be consistent with the return place before updating compiler-builtins, I wonder if the patched version isn't getting built for some reason.

bjorn3 reviewed (approved):
The cg_clif half LGTM. https://github.com/rust-lang/rust/pull/135328 will have two minor conflicts with this PR though.

bors commented:
:umbrella: The latest upstream changes (presumably #135328) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

tgross35 commented:
Sorry all, lost a rebase and I guess I'm a cloud VM (...?)

jieyouxu commented:
> Sorry all, lost a rebase and I guess I'm a cloud VM (...?)

Normal dev-desktop behavior

antoyo commented on compiler/rustc_codegen_gcc/src/int.rs:
```diff
@@ -364,80 +354,85 @@ impl<'a, 'gcc, 'tcx> Builder<'a, 'gcc, 'tcx> {
+
+        let a_elem_type = a_type.dyncast_array().expect("non-array a value");
+        debug_assert!(a_elem_type.is_integral());
+        let res_ty = match width {
+            32 => self.tcx.types.i32,
+            64 => self.tcx.types.i64,
+            128 => self.tcx.types.i128,
+            _ => unreachable!("unexpected integer size"),
+        };
```
Was there a reason to do that instead of reusing `a_type`?

tgross35 replied:
I believe `a_type` is the GCC type and I need the rustc type to calculate the layout. I could not find a way to convert in this direction - is there one? Or a way to get a layout directly from the GCC type (feel like I might be missing something here).

antoyo replied:
Oh, sorry I forgot that you now have multiple types to handle here and not only 128-bit integers.
It's good as it is.

antoyo commented on compiler/rustc_codegen_gcc/src/int.rs:
```diff
@@ -364,80 +354,85 @@ impl<'a, 'gcc, 'tcx> Builder<'a, 'gcc, 'tcx> {
+        let oflow_param_type = oflow_type.make_pointer();
+        let res_type = a_type;
+
+        let oflow_value = self.current_func().new_local(self.location, oflow_type, "overflow");
```
Please rename all `oflow` to `overflow` to make it a bit clearer.

tgross35 replied:
This is done

antoyo commented on compiler/rustc_codegen_gcc/build_system/build_sysroot/Cargo.toml:
```diff
@@ -13,6 +13,8 @@ test = { path = "./sysroot_src/library/test" }
 proc_macro = { path = "./sysroot_src/library/proc_macro" }
 
 [patch.crates-io]
+# todo: remove patch
+compiler_builtins = { git = "https://github.com/tgross35/compiler-builtins.git", package = "compiler_builtins", branch = "NOMERGE-overflowing-c-safe-ret-version" }
```
Only this to remove, I guess.

antoyo reviewed (approved):
This looks good to me.
Thanks a lot!

tgross35 commented:
Thanks for reviewing!

Actually merging this is still blocked until I can update compiler-builtins. Once that works again, I'll drop the Cargo patches, bump the version, and merge this.

@rustbot blocked

bors commented:
:umbrella: The latest upstream changes (presumably #135180) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

tgross35 committed:
Update compiler-builtins to 0.1.143

0.1.142 fixes an issue parsing optimization flags, and 0.1.143 changes
`__rust_[ui]128_*` builtins to use a C-safe signature.

tgross35 commented:
My diagnostics fix merged, which meant the r-l/rust update to builtins 0.1.141 could merge, which meant I could merge the signature changes to builtins, and now this is finally unblocked. Dropped the branch override patch and replaced it with a bump to builtins 0.1.143. Cranelift tests pass locally.

@bors r=bjorn3,antoyo

bors commented:
:pushpin: Commit f6a2db8e1b6ee82ebf44b82dfc83740051c59d02 has been approved by `bjorn3,antoyo`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=bjorn3,antoyo f6a2db8e1b6ee82ebf44b82dfc83740051c59d02 -->
<!-- homu: {"type":"Approved","sha":"f6a2db8e1b6ee82ebf44b82dfc83740051c59d02","approver":"bjorn3,antoyo","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
